executor/cook/executor.py

```diff
@@ -0,0 +1,384 @@
+        :param send_progress_update: The function to invoke while sending progress updates
+        :return: Nothing.
+    """
+    try:
+        task_id = get_task_id(task)
+        for current_progress in progress_watcher.retrieve_progress_states():
+            logging.debug('Latest progress: {}'.format(current_progress))
+            send_progress_update(driver, task_id, max_message_length, current_progress)
+    finally:
+        progress_complete_event.set()
+
+
+def launch_progress_tracker(driver, task, max_message_length, progress_watcher, progress_updater):
+    """
+        Launches the threads that track progress and send progress updates to the driver.
+        :return: Nothing
+    """
+    progress_complete_event = Event()
+    progress_update_thread = Thread(target=track_progress,
+                                    args=(driver, task, progress_watcher, max_message_length, progress_complete_event,
+                                          progress_updater.send_progress_update))
+    progress_update_thread.start()
+    return progress_complete_event
+
+
+class ProgressWatcher(object):
+    """
+        This class tails the output from the target file listening for progress messages.
+        The retrieve_progress_states generates all progress messages iteratively.
+    """
+    @staticmethod
+    def match_progress_update(progress_regex_str, input_string):
+        matches = re.findall(progress_regex_str, input_string)
+        return matches[0] if len(matches) >= 1 else None
+
+    def __init__(self, target_file, progress_regex_string, completed_signal):
+        self.target_file = target_file
+        self.progress_regex_string = progress_regex_string
+        self.progress = None
+        self.completed_signal = completed_signal
+
+    def current_progress(self):
+        return self.progress
+
+    def tail(self, sleep_time_ms):
+        """
+        This method incrementally generates lines from a file by waiting for new content from a file.
+        It behaves like the 'tail -f' shell command.
+        :param sleep_time_ms: The unit of time in ms to repetitively sleep when the file has not been created or no new
+          content is available in the file being tailed.
+        :return: an incrementally generated list of lines in the file being tailed.
+        """
+        while True:
+            if os.path.isfile(self.target_file):
+                logging.info('{} has been created, reading contents'.format(self.target_file))
+                break
+            if self.completed_signal.isSet():
+                logging.info('No output has been read to parse progress messages')
+                return
+            logging.debug('{} has not yet been created, sleeping {} ms'.format(self.target_file, sleep_time_ms))
+            time.sleep(sleep_time_ms / 1000.0)  # sleep briefly
+
+        target_file_obj = open(self.target_file, 'r')
+        while True:
```

Comment:
If I spam you with progress updates, you're going to build up a huge queue of lines to read.  Shouldn't we be ensuring we're reading to the last progress update line during each interval?